File: ArraysAndHashing/Java/EncodeAndDecodeStrings/Codec.java
/*
 * Time complexity: O(m), where:
 * - m is the amount of all characters in the strings combined
 * Space complexity: O(m + n)
 * - n is the n of strings in the input arr
 *
 *
 * The logic:
 * Doesn't matter how you incode the string, but use length of the word that
 * shows when the word starts and ends.
 * Decode func will use two pointers approach, where i is used to find the word
 * and j is for finding a delimeter, which is # here.
 * When the words are found, they are added to the list.
 *
 * Keynote: different types can be added to a StringBuilder
 * (Objects must be used with caution), that's why the error is not thrown when
 * adding an integer or a char to sb.
 * The same would happen with String (though the new String is created every time),
 * because String uses an array of characters or
 * StringBuilder underneath.
 *
 */

package Java.EncodeAndDecodeStrings;

import java.util.*;

public class Codec {
  public static String encode(String[] words) {
    StringBuilder sb = new StringBuilder();
    for (String word : words) { // O(m)
      int len = word.length();
      char delimeter = '#';
      sb.append(len).append(delimeter).append(word); // no type error
    }
    return sb.toString(); // could be new String(sb)
  }

  public static ArrayList<String> decode(String str) {
    ArrayList<String> res = new ArrayList<>();
    int i = 0;
    while (i < str.length()) { // O(m)
      int j = i;
      // look for delimeter
      while (str.charAt(j) != '#') { // O(m)
        j++;
      }

      int len = Integer.parseInt(str.substring(i, j)); // O(m)
      // find when the word starts
      i = j + 1;
      res.add(str.substring(i, i + len)); // O(m)
      // find the next word
      i = i + len;
    }
    return res;
  }

  private static void printList(ArrayList<String> words) {
    for (String word : words) {
      System.out.println(word);
    }
  }

  public static void main(String[] args) {
    String[] words = { "hello", "world" };
    String encodedStr = encode(words);
    System.out.println(encodedStr); // 5#hello5#world

    ArrayList<String> res = decode(encodedStr);
    printList(res); // hello world

  }
}

File: ArraysAndHashing/Python/EncodeAndDecodeStrings/codec.py
'''
 * Time complexity: O(m), where:
 * - m is the amount of all characters in the strings combined
 * Space complexity: O(m + n)
 * - n is the n of strings in the input arr
 *
 * The logic:
 * Doesn't matter how you incode the string, but use length of the word that
 * shows when the word starts and ends.
 * Decode func will use two pointers approach, where i is used to find the word
 * and j is for finding a delimeter, which is # here.
 * When the words are found, they are added to the list.
 *
 '''

from typing import List

class Solution:
    def encode(self, strs: List[str]) -> str:
      res = ""
      for s in strs:
        res += str(len(s)) + "#" + s
      return res

    def decode(self, s: str) -> List[str]:
      res = []
      i = 0

      while i < len(s):
        j = i
        while (s[j] != "#"):
          j += 1

        length = int(s[i:j])
        i = j + 1
        res.append(s[i:i+length])
        i = i + length

      return res


if __name__ == "__main__":
  s = Solution()
  words = ["hello", "world"]
  string = s.encode(words)
  print(string) # 5#hello5#world

  print(s.decode(string)) # ['hello', 'world']
